backend/src/Edna/ExperimentReader/Parser.hs

```diff
@@ -0,0 +1,123 @@
+module Edna.ExperimentReader.Parser
+  ( parseExperimentXls
+  , processWorkSheet
+  , ParserType
+  ) where
+
+import Universum
+
+import qualified Data.ByteString.Lazy as L
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map as M
+
+import Codec.Xlsx (CellValue(..), Worksheet(..), cellValue, ixCell, toXlsxEither, wsCells, xlSheets)
+import Data.List.Split (wordsBy)
+import Data.Text (replace)
+import Fmt (pretty)
+import Lens.Micro (ix, _Just)
+import Text.Read (readParen)
+
+import Edna.ExperimentReader.Error (ExperimentParsingError(..))
+import Edna.ExperimentReader.Types
+  (CellType(..), Parameter(..), ParameterType(..), PointYX(..), Signal(..), PlateUnit(..))
+import Edna.Web.Types (ExperimentalMeasurement(..))
+
+type ParserType a = Either ExperimentParsingError a
+
+-- | Based on specific CellType return its value or get and exception that cell is empty or its
+-- expected type does not satisfy actual type
+specificCellAt
+  :: Typeable a
+  => Worksheet
+  -> PointYX
+  -> CellType a
+  -> Either ExperimentParsingError a
+specificCellAt workSheet p@(PointYX position) cellType =
+  let cell = workSheet ^? ixCell position . cellValue . _Just in
+  case (cellType, cell) of
+    (_, Nothing) -> Left EmptyCell
+    (_, Just (CellText "")) -> Left EmptyCell
+
+    (CText, Just (CellText x)) -> Right x
+    (CDouble, Just (CellDouble x)) -> Right x
+
+    (CSignal, Just (CellDouble x)) -> Right $ Signal x False
+    (CSignal, Just (CellText
+      -- If signal value is in brackets then it is outlier
+      (readParen True reads . toString . replace "," "." -> ((d, "") : _))
+      )) -> Right $ Signal d True
+
+    (_, cv) -> Left $ UnexpectedCellType p cellType cv
+
+-- | Check that cell representation equals to some value
+cellSatisfy :: Worksheet -> Text -> PointYX -> Bool
+cellSatisfy workSheet value (PointYX position) = case workSheet ^? ixCell position . cellValue of
+  Just (Just x) -> pretty x == value
```

Comment:
I don't like it because it will work weirdly if a cell contains `CellRich` or `CellError` for example. AFAIU we always expect text here, so maybe let's do `== value` only when we encounter `CellText` and return `False` for all other `CellValue` constructors? Like, if we pass `Text`, we probably expect `CellText`. If we want to support other types of cells, we should pass `CellValue` here (or `Maybe CellValue`).

Reply:
>we always expect text here

No, it checks that cell text representation is the same as we want

Reply:
I mean, this function should always fail when it encounters cell type other than `CellText`, no? Because we pass only empty string and `<>` IIRC.

Reply:
No, why?\
Oh, wait, I now understand what you mean. You mean that I am using it only for comparing `CellText` values, right? That make sense, but initially it was made to compare for example not only `CellText` values but also `"1,1 == CellDouble 1,1"`\
So, if you want I can remove it

Reply:
> You mean that I am using it only for comparing `CellText` values, right?

Yes.

> but initially it was made to compare for example not only `CellText` values but also `"1,1" == CellDouble 1,1`

Yes, and I don't like it because it complicates semantic of this function because now one needs to think whether `CellText "1,1"` is treated the same way as `CellDouble 1.1` So if you want it to be general, I think it's better to pass `CellValue` rather than just `Text`. But in this case it doesn't need to be general, so I would simply restrict it to `CellText` values.